modules/renter/streamer.go

```diff
@@ -0,0 +1,141 @@
+		file: file,
+		r:    r,
+	}
+	return file.name, s, nil
+}
+
+// Read implements the standard Read interface. It will download the requested
+// data from the sia network and block until the download is complete.  To
+// prevent http.ServeContent from requesting too much data at once, Read can
+// only request a single chunk at once.
+func (s *streamer) Read(p []byte) (n int, err error) {
+	// Get the file's size
+	s.file.mu.RLock()
+	fileSize := int64(s.file.size)
+	s.file.mu.RUnlock()
+
+	// Make sure we haven't reached the EOF yet.
+	if s.offset >= fileSize {
+		return 0, io.EOF
+	}
+
+	// Calculate how much we can download. We never download more than a single chunk.
+	chunkSize := s.file.staticChunkSize()
+	remainingData := uint64(fileSize - s.offset)
+	requestedData := uint64(len(p))
+	remainingChunk := chunkSize - uint64(s.offset)%chunkSize
+	length := min(chunkSize, remainingData, requestedData, remainingChunk)
+
+	// Download data
+	buffer := bytes.NewBuffer([]byte{})
+	d, err := s.r.newDownload(downloadParams{
+		destination:       newDownloadDestinationWriteCloserFromWriter(buffer),
+		destinationType:   destinationTypeSeekStream,
+		destinationString: "httpresponse",
+		file:              s.file,
+
+		latencyTarget: 25e3 * time.Millisecond, // TODO high default until full latency suport is added.
+		length:        length,
+		needsMemory:   true,
+		offset:        uint64(s.offset),
+		overdrive:     3,  // TODO: moderate default until full overdrive support is added.
+		priority:      10, // TODO: high default until full priority support is added.
+	})
+	if err != nil {
+		return 0, errors.AddContext(err, "failed to create new download")
+	}
+
+	// Add the download object to the download queue.
+	// TODO: Maybe this is not necessary for streams?
+	s.r.downloadHistoryMu.Lock()
+	s.r.downloadHistory = append(s.r.downloadHistory, d)
+	s.r.downloadHistoryMu.Unlock()
+
+	// Block until the download has completed.
+	select {
+	case <-d.completeChan:
+		if d.Err() != nil {
+			return 0, errors.AddContext(d.Err(), "download failed")
+		}
+	case <-s.r.tg.StopChan():
+		return 0, errors.New("download interrupted by shutdown")
+	}
+
+	// Copy downloaded data into buffer.
+	copy(p, buffer.Bytes())
+
+	// Adjust offset
+	s.offset += int64(length)
+	return int(length), nil
+}
+
+// Seek sets the offset for the next Read to offset, interpreted
+// according to whence: SeekStart means relative to the start of the file,
+// SeekCurrent means relative to the current offset, and SeekEnd means relative
+// to the end. Seek returns the new offset relative to the start of the file
+// and an error, if any.
+func (s *streamer) Seek(offset int64, whence int) (int64, error) {
+	var newOffset int64
+	switch whence {
+	case io.SeekStart:
+		newOffset = 0
+	case io.SeekCurrent:
+		newOffset = s.offset
+	case io.SeekEnd:
+		s.file.mu.RLock()
+		newOffset = int64(s.file.size)
+		s.file.mu.RUnlock()
+	}
+	newOffset += offset
+
+	if newOffset < 0 {
```

Comment:
should we be checking for overflow here as well?

Reply:
file.Seek I believe won't return an error, but that's because you can call `file.Write` at the end of a file even if you are past the last piece of already written data.

@lukechampine do you have thoughts here?

Reply:
The `io.Seeker` docstring says:

>Seeking to an offset before the start of the file is an error. Seeking to any positive offset is legal, but the behavior of subsequent I/O operations on the underlying object is implementation-dependent. 

The standard library seems to be split on this. When seeking past the end of the stream, `strings.Reader` and `bytes.Reader` return `nil`, but future `Read` calls return `io.EOF`. On the other hand, `os.File` (and the more obscure `elf.readSeekerFromReader`) return an error immediately from `Seek`.

I'm slightly in favor of returning an error immediately, because seeking past the end does feel like an invalid operation that ought to return an error. The only downside that I can see is that people may be less likely to check `Seek` errors than `Read` errors. But any code that does responsible error checking should work fine with either approach.

Reply:
@lukechampine yeah that's what I did to decide on how to implement it. I first tried doing the same on an `os.File`.